### PR TITLE
Move CI provisioning from v1 workspaces to v2 clusters

### PR DESCRIPTION
**Pull request by kesmit13**
Retargets the CI provisioning scripts and workflows from the v1
workspace-group flow at v2 clusters, which is what `management.version`
already defaults to.

## Provisioning

`resources/create_test_cluster.py` is now a single `manage_clusters()` /
`create_cluster()` call. `wait_on_active=True` covers ACTIVE, then the
endpoint, then the firewall, which replaces both polling loops and a bare
`time.sleep(10)`. `--expires` was parsed and then never used, so CI
clusters had no expiry at all; it is wired to `expires_at` now.
`resources/drop_test_cluster.py` takes a cluster ID and terminates it.

The admin password needs care. `POST /v2/clusters` accepts an
`adminPassword` and ignores it, generating its own, and `PATCH` does the
same, so the password can only be read off the create response and is
`None` after any refresh. That value cannot reach a consuming job: it has
to be masked, and the runner drops any output whose value matches a mask
-- "Skip output 'cluster-password' since it may contain secret" -- so
masking it and passing it between jobs are mutually exclusive.

So the generated password is traded for `secrets.CLUSTER_PASSWORD` over
SQL, once, while the script still holds it: `ALTER USER 'admin'@'%'
IDENTIFIED BY ...` (`SET PASSWORD` wants a pre-hashed value and rejects a
literal with `1372: Password hash should be a 41-digit hexadecimal
number`). No password is reported at all after that, and every consuming
job reads the secret it already had. Keeping the credential a secret
rather than a job output is also why `SINGLESTOREDB_URL` interpolates it
raw: a percent-encoded copy is a different string from the mask, so
emitting one would put a live credential past the runner's masking. The
secret is maintainer-set, so keeping it alphanumeric is a cheaper
constraint than that; both consuming sites say so.

The ID, host, port and database name are reported as soon as
`create_cluster()` returns, before that password reset and before the SQL
load. Both can fail against a cluster that is already running and already
billing, and a teardown job with an empty `cluster-id` would send its
DELETE to `/v2/clusters/` and leak it. The two shutdown steps refuse an
empty ID outright now.

`secrets.CLUSTER_USER` and `vars.CLUSTER_PROJECT` are gone. A cluster has
exactly one user, `admin`, and no route creates another, so the secret
could only ever hold that value; the project is named in the workflow so
the deployment target is visible next to the create call. Once this
merges, `CLUSTER_USER` can be deleted from repository settings.

## A datetime bug this uncovered

`expires_at` read back as `None` on a cluster that demonstrably had an
expiry set. `expiresAt` is the one field the API returns in Go's
`time.Time.String()` format -- `2026-09-17 14:42:41.445984 +0000 UTC` --
while every other timestamp is RFC 3339. `to_datetime` split on `.` to
pad the fraction, produced `445984 +0000 UTC`, and parsing returned
`None`. Pre-existing, and it also left the expiry column blank in
`SHOW CLUSTERS`. Fixed with tests for both shapes.

Accepting that shape then exposed its zero value: an unset Go
`time.Time` renders as `0001-01-01 00:00:00 +0000 UTC`, and only the
RFC 3339 spelling of that sentinel was recognized, so a resource that
does not expire reported an expiry in year 1. Both `to_datetime` and
`to_datetime_strict` now test the parsed value for January 1 of year 1,
which covers every spelling, and do it before the UTC shift, which can
fall below `MINYEAR` on a year-1 value.

## Workflow linting

There was no linter over `.github/workflows/`, so nothing checked
expression syntax, `needs.*.outputs.*` references or the shell in `run:`
blocks. `actionlint` is now a pre-commit hook -- a pip wrapper, so no Go
toolchain or Docker -- and the 17 pre-existing findings it reported are
fixed: stale action majors, a step named after `${{
matrix.python-version }}` in a job whose matrix only defines `os`, and
six unquoted `$GITHUB_OUTPUT` redirects.

Every action pin is also off the Node.js 20 runtime now. The first pass
bumped to `checkout@v4` / `setup-python@v5`, which satisfied actionlint
but still declared node20; the pins are the lowest major of each action
that declares node24, read from `action.yml` at the tag. Verified by the
run annotations disappearing between 6c22959d and 88283263.

## Verified

- pre-commit clean, including actionlint on all six workflows, on 3.9-3.13
- live create: ACTIVE in 5:54, `test.sql` loaded, `expires_at` exactly
  1h after `created_at`
- `test_basics` against that cluster: 30 passed, 2 skipped
- live teardown both ways: `drop_test_cluster.py` to TERMINATED in 2.8s,
  and the workflow's `curl DELETE /v2/clusters/{id}?force=true` returning
  200. No clusters leaked
- marker partition exact: 905 selected + 65 deselected = 970
- the zero-time sentinel in both shapes and both helpers, by unit test

## Not verified

The reordered reporting block is not covered by a test, and the live
create above predates it. The `publish.yml` artifact-action bumps only
execute on a tag or release, so they land untested here.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

---

> [!NOTE]
> **High Risk**
> Changes live cloud provisioning, teardown, and cancellation cleanup paths that directly affect billing and org cluster quota; datetime and DELETE parameter fixes touch production management client behavior.
> 
> **Overview**
> Retargets CI cluster setup from **v1 workspace groups** to **v2 clusters** (`create_test_cluster.py`, `drop_test_cluster.py`, publish/smoke shutdown via `DELETE /v2/clusters/{id}?force=true`), fixes admin auth by resetting the API-generated password with `ALTER USER` to `CLUSTER_PASSWORD`, and standardizes on the **`admin`** user instead of `CLUSTER_USER`.
> 
> Adds a **JSONL deployment ledger** (`SINGLESTOREDB_TEST_DEPLOYMENT_LOG`) written on each management create, plus an **`if: always()`** cleanup step that runs `cleanup_deployments --ledger` so cancelled jobs still terminate billable clusters. Related test harness changes: bounded **terminate retries** for provisioning-state 4xx, xdist worker reporting of stranded deployments, safer fusion/v1 teardowns, and **`-m 'not management_v1'`** on PR checks with v1-only tests moved to a nightly **`coverage.yml`** job.
> 
> Fixes **`to_datetime`** for Go `time.Time` strings and zero-time sentinels; sends **`force` as `'true'/'false'`** on management DELETEs. CI hygiene: **workflow concurrency** on `code-check`, **full git history** for the change detector, **actionlint** pre-commit, action pin bumps, and default **pytest-xdist workers 3→2**.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit caed76451d1e3ce9ec09ab41bc2b8eb5bddc47bb. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: .github/workflows/publish.yml [hunk 1]
```diff
@@ -118,7 +131,10 @@ jobs:
           # points --pyargs at the workspace, so that pyproject is the inifile
           # here. Without the plugin pytest exits on the unknown arguments.
           CIBW_TEST_REQUIRES: "pytest pytest-xdist"
-          CIBW_ENVIRONMENT: "SINGLESTOREDB_URL='mysql://${{ secrets.CLUSTER_USER }}:${{ secrets.CLUSTER_PASSWORD }}@${{ needs.setup-database.outputs.cluster-host }}:3306/${{ needs.setup-database.outputs.cluster-database }}?pure_python=0'"
+          # CLUSTER_PASSWORD has to survive both the userinfo half of the URL
+          # and the single-quoted shell word cibuildwheel evaluates, so keep the
+          # secret alphanumeric: no ':', '@', '/', '%' or quote characters.
+          CIBW_ENVIRONMENT: "SINGLESTOREDB_URL='mysql://admin:${{ secrets.CLUSTER_PASSWORD }}@${{ needs.setup-database.outputs.cluster-host }}:3306/${{ needs.setup-database.outputs.cluster-database }}?pure_python=0'"
```
[the remark this answers is not in the document]
Comment (kesmit13) [reply .github/workflows/publish.yml:137]: Declined, deliberately.

`CLUSTER_PASSWORD` is a repository secret this repo's maintainers set, not an API-generated value. That was the point of 855a593c: the generated password *is* drawn from the full printable set (one observed value was `{:D}TK*[F3Ll}Ups2pNv`), which is exactly why it is no longer propagated — the script trades it for the known secret over `ALTER USER` while it still holds it. So the character set here is ours to choose, and the restriction is documented at both sites that consume it.

Encoding it is not free. The percent-encoded form is a different string from the mask, so emitting it — which is what the original design did, as `cluster-password-url` — puts a live credential past the runner's masking and into a job output. Keeping the raw secret in a masked env var and accepting an alphanumeric-secret constraint is the better trade of the two.

Separate connection variables are not available either: `singlestoredb/tests/conftest.py` starts a Docker container whenever `SINGLESTOREDB_URL` is unset, so the URL is what selects a remote cluster at all — `SINGLESTOREDB_HOST`/`_PASSWORD` would silently test against a local container instead.

File: .github/workflows/smoke-test.yml [hunk 2]
```diff
@@ -118,7 +135,10 @@ jobs:
         run: pytest -v --pyargs singlestoredb.tests.test_basics
         env:
           PYTHONPATH: ${{ github.workspace }}
-          SINGLESTOREDB_URL: "${{ matrix.driver }}://${{ secrets.CLUSTER_USER }}:${{ secrets.CLUSTER_PASSWORD }}@${{ needs.setup-database.outputs.cluster-host }}:3306/${{ needs.setup-database.outputs.cluster-database }}?pure_python=${{ matrix.pure-python }}&buffered=${{ matrix.buffered }}"
+          # CLUSTER_PASSWORD goes into the userinfo half of a URL here, so it
+          # has to be free of characters that would need percent-encoding --
+          # ':', '@', '/', '%' and the like. Keep the secret alphanumeric.
+          SINGLESTOREDB_URL: "${{ matrix.driver }}://admin:${{ secrets.CLUSTER_PASSWORD }}@${{ needs.setup-database.outputs.cluster-host }}:3306/${{ needs.setup-database.outputs.cluster-database }}?pure_python=${{ matrix.pure-python }}&buffered=${{ matrix.buffered }}"
```
[the remark this answers is not in the document]
Comment (kesmit13) [reply .github/workflows/smoke-test.yml:141]: Declined for the same reason as https://github.com/singlestore-labs/singlestoredb-python/pull/134#discussion_r4049397781, which covers both this line and line 154: `CLUSTER_PASSWORD` is a maintainer-set repository secret rather than the API-generated value, so its character set is ours to pick; percent-encoding it would mean emitting a string that does not match the mask; and the URL is what keeps the tests off a local Docker container, so per-field connection variables are not an option here.

File: resources/create_test_cluster.py [hunk 3]
```diff
@@ -66,120 +72,170 @@
...
 if options.output == 'env':
-    print(f'CLUSTER_ID={ws.id}')
+    print(f'CLUSTER_ID={cluster.id}')
     print(f'CLUSTER_HOST={host}')
     print(f'CLUSTER_PORT={port}')
     print(f'CLUSTER_DATABASE={database}')
 elif options.output == 'github':
     with open(os.environ['GITHUB_OUTPUT'], 'a') as output:
-        print(f'cluster-id={ws.id}', file=output)
+        print(f'cluster-id={cluster.id}', file=output)
```
[the remark this answers is not in the document]
Comment (kesmit13) [reply resources/create_test_cluster.py]: Fixed in 4e49cfa5.

The reporting block now runs as soon as `create_cluster()` returns. Everything it prints is already known there — `cluster.id`, the host/port split off `cluster.endpoint`, and the database name, which is generated locally — so the generated-password check, the `ALTER USER` and the SQL load can no longer fail with a live cluster and no ID anywhere.

The two shutdown steps also refuse an empty ID outright rather than requesting `/v2/clusters/?force=true` and moving on, and `curl --fail-with-body` makes a rejected DELETE fail the step instead of printing the error and exiting 0.

File: singlestoredb/management/utils.py [hunk 4]
```diff
@@ -418,18 +507,14 @@ def to_datetime(
...
     if obj == '0001-01-01T00:00:00Z':
         return None
-    obj = obj.replace('Z', '')
-    # Fix datetimes with truncated zeros
-    if '.' in obj:
-        obj, micros = obj.split('.', 1)
-        micros = micros + '0' * (6 - len(micros))
-        obj = obj + '.' + micros
-    out = converters.datetime_fromisoformat(obj)
+    out = converters.datetime_fromisoformat(_normalize_datetime(obj))
```
[the remark this answers is not in the document]
Comment (kesmit13) [reply singlestoredb/management/utils.py:533]: Good catch, and reproduced — the two spellings disagreed:

```
'0001-01-01 00:00:00 +0000 UTC' -> datetime.datetime(1, 1, 1, 0, 0)
'0001-01-01T00:00:00Z'          -> None
```

Fixed in 4e49cfa5. Rather than add a second string literal, both helpers now test the *parsed* value for Go's zero time (January 1 of year 1) via one shared `_is_go_zero_time` predicate, which covers every spelling — offset, zone name and monotonic reading included. The check runs before `_as_naive_utc`, since shifting an aware year-1 value onto UTC can carry it below `MINYEAR` and raise rather than returning the `None` the value means.

Regression tests added for both helpers: the Go spelling, a monotonic-reading variant, a non-UTC zone name and the date-only form all return `None`, and `to_datetime_strict` raises `ValueError` on the Go spelling.

File: resources/create_test_cluster.py [hunk 5]
```diff
@@ -66,120 +72,170 @@
...
+# control plane leaves the new password alone afterwards.
+password = options.password
+escaped = password.replace('\\', '\\\\').replace("'", "\\'")
+
+with s2.connect(
+    host=host, port=port, user='admin',
+    password=generated, connect_timeout=30,
+) as conn:
+    with conn.cursor() as cur:
+        cur.execute(f"ALTER USER 'admin'@'%' IDENTIFIED BY '{escaped}'")
```
[the remark this answers is not in the document]
Comment (kesmit13) [reply resources/create_test_cluster.py:236]: You are reading a stale description rather than stale code.

The `cluster-password` / `cluster-password-url` outputs were the original design in 397b4418. 855a593c replaced them with the `ALTER USER` reset, because the runner refuses to write a job output whose value matches a mask — "Skip output 'cluster-password' since it may contain secret" — which makes "masked" and "passed between jobs" mutually exclusive for the generated password. The description was never rewritten to match.

It has been now: the credential section describes the `ALTER USER` trade, and the `Verified` / `Not verified` sections are corrected along with it.

File: singlestoredb/tests/cleanup_deployments.py [hunk 6]
```diff
@@ -304,12 +314,297 @@ def keep(obj: Any) -> bool:
...
+    if not yes:
+        print('\nDry run; pass --yes to terminate these.')
+        return 0
+
+    from singlestoredb.tests import utils
+
+    failed = 0
+    for label, obj in leftovers:
+        try:
+            utils.terminate(obj)
```
[the remark this answers is not in the document]
Comment (kesmit13) [reply singlestoredb/tests/cleanup_deployments.py]: Fixed. You're right on both halves: the reaper is the end of the line, so it had no business borrowing the budget that exists to keep the per-class sweep from stalling the suite. `cleanup_deployments.py` now has its own `TERMINATE_TIMEOUT = 600.0`, passed at both sweep call sites (the ledger mode and the age/pattern mode — the latter can also be handed a provisioning cluster via `--since`), which covers the ~460s worst-case S-00 provision. `utils.TERMINATE_RETRY_TIMEOUT`'s comment no longer claims a handoff that wasn't happening. Added a test that pins the reaper's budget above the per-class one so the two can't drift back together. (addressed in 44c8512d754b66f8232fab1bab1e2bfadc613ce2)

File: .github/workflows/code-check.yml [hunk 7]
```diff
@@ -8,6 +8,19 @@ on:
...
+# one rather than letting both provision. The management step deploys ~6
+# clusters and holds them for the length of the suite, so two overlapping runs
+# double what the org is carrying for no extra coverage. Cancellation is safe
+# here because the cleanup step is `if: always()` and reaps from the ledger.
+#
+# Not applied to coverage.yml: it is schedule/dispatch only, and its two jobs
+# are meant to run at once with separate ledgers.
+concurrency:
+  group: ${{ github.workflow }}-${{ github.ref }}
+  cancel-in-progress: true
```
[the remark this answers is not in the document]
Comment (kesmit13) [reply .github/workflows/code-check.yml]: Good catch, and the premise checks out — GitHub puts the server's cancellation timeout at 5 minutes, after which remaining steps are force-terminated whether or not they are `if: always()`. So a 600s budget is unspendable in exactly the case it was sized for: an S-00 `DELETE` is refused until ACTIVE (~460s), and the step is dead before that.

Dropped `cancel-in-progress` from code-check.yml. A run that is not cancelled reaches the reaper with the job's full budget, which is where the 600s is real. Letting both runs finish costs clusters; cancelling them costs *stranded* clusters, which is worse.

Being straight about what this does not fix: a manually cancelled run can still be killed mid-provision, and there is no scheduled janitor, so that remainder needs `cleanup_deployments.py --older-than` run by hand. The comments on both workflows' cleanup steps and on `TERMINATE_TIMEOUT` now say that instead of implying `always()` is a guarantee. Incident 35631802648 stays covered either way — it was cancelled 19 minutes in, well past ACTIVE. (addressed in a981afefc6f12935f4a5c3c64986768101709047)

File: .github/workflows/coverage.yml [hunk 8]
```diff
@@ -82,3 +92,108 @@ jobs:
...
+    # Waits for test-coverage rather than running alongside it, so the v1
+    # workspace groups are never in flight at the same time as the v2 suite's
+    # cluster pool -- together they put more on the org than it wants to carry.
+    # This is a nightly cron, so the extra wall clock costs nothing.
+    #
+    # if: always() because this is a legacy gate, not a downstream build: a v2
+    # failure above says nothing about the v1 endpoints, and skipping v1 for it
+    # would hide a v1 regression behind an unrelated one.
+    needs: test-coverage
+    if: always()
```
[the remark this answers is not in the document]
Comment (kesmit13) [reply .github/workflows/coverage.yml]: Right — `always()` is broader than what the comment claims the intent is. Switched to `if: ${{ !cancelled() }}`: the v1 gate still runs when the v2 job fails (a v2 failure says nothing about the v1 endpoints), but a cancelled nightly no longer goes on to deploy workspace groups whose `always()` cleanup step would be force-killed inside GitHub's 5-minute cancellation window — the same window as the `cancel-in-progress` finding above. The comment now says that rather than just "legacy gate". (addressed in 28153d76)

File: .github/workflows/code-check.yml (hunk 7 above)
[the remark this answers is not in the document]
Comment (kesmit13) [reply .github/workflows/code-check.yml]: Already addressed in a981afef, which dropped `cancel-in-progress` from `code-check.yml` entirely — see the reply above for the reasoning. Leaving this open for you to resolve.

Worth noting one follow-on from this finding, since the five-minute window is the premise of both: every management fixture now passes `expires_at` (cb13ea0e). The sweep, the ledger and `tearDownClass` are all in-process, so none of them survives the runner being force-killed, which is exactly the case this comment identified. An expiry is a property of the deployment, so the control plane honours it either way — it puts a ceiling on what a killed run can strand, rather than relying on cleanup that may never get to run.
